diff --git a/cpp/src/arrow/BUILD.bazel b/cpp/src/arrow/BUILD.bazel
new file mode 100644
index 0000000000..e3302eaf1c
--- /dev/null
+++ b/cpp/src/arrow/BUILD.bazel
@@ -0,0 +1,225 @@
+package(default_visibility = ["//visibility:public"])
+
+# Generate the header file, util/config.h
+genrule(
+    name = "util_config_h",
+    srcs = ["util/config.h.cmake"],
+    outs = ["util/config.h"],
+    cmd = """
+        # Start with the source file
+        cat $(location util/config.h.cmake) > $@
+
+        # Perform @VAR@ substitutions
+        sed \
+            -e 's/@ARROW_VERSION_MAJOR@/21/g' \
+            -e 's/@ARROW_VERSION_MINOR@/0/g' \
+            -e 's/@ARROW_VERSION_PATCH@/0/g' \
+            -e 's/@ARROW_VERSION@/21.0.0/g' \
+            -e 's/@ARROW_SO_VERSION@/2100/g' \
+            -e 's/@ARROW_FULL_SO_VERSION@/2100.0.0/g' \
+            -e 's/@ARROW_CXX_COMPILER_ID@/GNU/g' \
+            -e 's/@ARROW_CXX_COMPILER_VERSION@/13.3.0/g' \
+            -e 's/@CMAKE_CXX_FLAGS@/-O2/g' \
+            -e 's/@UPPERCASE_BUILD_TYPE@/RELEASE/g' \
+            -e 's/@ARROW_PACKAGE_KIND@/dev/g' \
+            $< >$@
+
+        # Perform #cmakedefine substitutions
+        # Use native int128 instead of boost::multi_precision
+        sed -i'' -e 's/#cmakedefine ARROW_USE_NATIVE_INT128/#define ARROW_USE_NATIVE_INT128/g' $@
+        sed -i'' -e 's/#cmakedefine ARROW_WITH_OPENTELEMETRY/#define ARROW_WITH_OPENTELEMETRY/g' $@
+        # Use jemalloc but not vendored
+        sed -i'' -e 's/#cmakedefine ARROW_JEMALLOC_VENDORED/#undef ARROW_JEMALLOC_VENDORED/g' $@
+        sed -i'' -e 's/#cmakedefine ARROW_JEMALLOC/#define ARROW_JEMALLOC/g' $@
+        sed -i'' -e 's/#cmakedefine/#undef/g' $@
+    """,
+)
+
+# Generate the header file, util/config_internal.h
+genrule(
+    name = "util_config_internal_h",
+    srcs = ["util/config_internal.h.cmake"],
+    outs = ["util/config_internal.h"],
+    cmd = """
+        # Start with the source file
+        cat $(location util/config_internal.h.cmake) > $@
+
+        # Perform @VAR@ substitutions
+        sed \
+            -e 's/@ARROW_GIT_ID@/ee4d09ebef/g' \
+            -e 's/@ARROW_GIT_DESCRIPTION@/apache-arrow-21.0.0-rc6/g' \
+            $< >$@
+    """,
+)
+
+cc_library(
+    name = "core",
+    srcs = glob(
+        [
+            "*.cc",
+            "array/*.cc",
+            "c/*.cc",
+            "compute/**/*.cc",
+            "extension/*.cc",
+            "io/*.cc",
+            "json/*.cc",
+            "telemetry/*.cc",
+            "tensor/*.cc",
+            "util/*.cc",
+            "vendored/**/*.cpp",
+            "vendored/**/*.cc",
+            # "testing/*.cc",
+        ],
+        exclude = [
+            "**/*_test.cc",
+            "**/test_*.cc",
+            "util/*_test.cc",
+            # "util/*simd*.cc",
+            "**/*avx*.cc", "**/*512*.cc",
+            "**/*_benchmark.cc",
+            "**/benchmark_*.cc",
+            "io/hdfs*.cc",
+            "vendored/datetime/tz.cpp",
+        ],
+    ),  #  + ["compute/test_util_internal.cc"],
+    hdrs = glob(
+        [
+            "*.h",
+            "array/*.h",
+            "c/*.h",
+            "compute/**/*.h",
+            "compute/kernels/*.inc.cc",
+            "extension/*.h",
+            "io/*.h",
+            "json/*.h",
+            "telemetry/*.h",
+            "tensor/*.h",
+            "util/*.h",
+            "vendored/**/*.h",
+            "vendored/**/*.hpp",
+            "vendored/datetime/tz.cpp",  # being included like a header file
+            # "testing/*.h",
+        ],
+        exclude = [
+            "**/*_test.h",
+            "**/test_*.h",
+            # "util/*simd*.h",
+            "**/*avx*.h", "**/*512*.h",
+            "io/hdfs*.h",
+        ],
+    ) + [
+        ":util_config_h",
+        ":util_config_internal_h",
+    ],  #  + ["compute/test_util_internal.h"],
+    copts = [
+        # "-DARROW_HAVE_RUNTIME_BMI2",
+        # "-mbmi2",
+        # "-DARROW_HAVE_RUNTIME_AVX2",
+        # "-mavx2",
+        # "-DARROW_HAVE_RUNTIME_AVX512",
+        # "-mavx512f",
+        # "-mavx512cd",
+        # "-mavx512bw",
+        # "-mavx512dq",
+        # "-mavx512vl",
+    ],
+    include_prefix = "arrow",
+    strip_include_prefix = "",
+    deps = [
+        "@jemalloc",
+        "@opentelemetry-cpp//exporters/ostream:ostream_log_record_exporter",
+        "@opentelemetry-cpp//exporters/ostream:ostream_span_exporter",
+        "@opentelemetry-cpp//exporters/otlp:otlp_http_exporter",
+        "@opentelemetry-cpp//exporters/otlp:otlp_http_log_record_exporter",
+        "@opentelemetry-proto//:trace_service_proto_cc",
+        "@protobuf",
+        "@rapidjson",
+        "@snappy", "@brotli", "@bzip2", "@zlib", "@zstd",
+        "@xsimd", "@lz4//:lz4_frame",
+    ],
+)
+
+cc_library(
+    name = "base",
+    srcs = glob(
+        [
+            "*.cc",
+            # "array/*.cc",
+            # "c/*.cc",
+            # "compute/**/*.cc",
+            # "extension/*.cc",
+            # "io/*.cc",
+            # "json/*.cc",
+            # "telemetry/*.cc",
+            # "tensor/*.cc",
+            # "util/*.cc",
+            # "vendored/**/*.cpp",
+            # "vendored/**/*.cc",
+        ],
+        exclude = [
+            "**/*_test.cc",
+            "**/test_*.cc",
+            # "util/*_test.cc",
+            # "util/*simd512*.cc",
+            # "util/*avx512*.cc",
+            # "**/*_benchmark.cc",
+            # "**/benchmark_*.cc",
+            # "io/hdfs*.cc",
+            # "vendored/datetime/tz.cpp",
+        ],
+    ),
+    hdrs = glob(
+        [
+            "*.h",
+            # "array/*.h",
+            # "c/*.h",
+            # "compute/**/*.h",
+            # "compute/kernels/*.inc.cc",
+            # "extension/*.h",
+            # "io/*.h",
+            # "json/*.h",
+            # "telemetry/*.h",
+            # "tensor/*.h",
+            # "util/*.h",
+            # "vendored/**/*.h",
+            # "vendored/**/*.hpp",
+            # "vendored/datetime/tz.cpp", # being included like a header file
+        ],
+        exclude = [
+            "**/*_test.h",
+            "**/test_*.h",
+            # "util/*simd512*.h",
+            # "util/*avx512*.h",
+            # "io/hdfs*.h",
+        ],
+    ) + [
+        ":util_config_h",
+        ":util_config_internal_h",
+    ],
+    copts = [
+        # "-DARROW_HAVE_RUNTIME_BMI2",
+        # "-mbmi2",
+        # "-DARROW_HAVE_RUNTIME_AVX2",
+        # "-mavx2",
+        # "-DARROW_HAVE_RUNTIME_AVX512",
+        # "-mavx512f",
+        # "-mavx512cd",
+        # "-mavx512bw",
+        # "-mavx512dq",
+        # "-mavx512vl",
+    ],
+    include_prefix = "arrow",
+    strip_include_prefix = "",
+    deps = [
+        "@jemalloc",
+        "@opentelemetry-cpp//exporters/ostream:ostream_log_record_exporter",
+        "@opentelemetry-cpp//exporters/ostream:ostream_span_exporter",
+        "@opentelemetry-cpp//exporters/otlp:otlp_http_exporter",
+        "@opentelemetry-cpp//exporters/otlp:otlp_http_log_record_exporter",
+        "@opentelemetry-proto//:trace_service_proto_cc",
+        "@protobuf",
+        "@rapidjson",
+        "@snappy",
+        "@xsimd",
+    ],
+)
diff --git a/cpp/src/arrow/acero/BUILD.bazel b/cpp/src/arrow/acero/BUILD.bazel
new file mode 100644
index 0000000000..48186e1f89
--- /dev/null
+++ b/cpp/src/arrow/acero/BUILD.bazel
@@ -0,0 +1,50 @@
+package(default_visibility = ["//visibility:public"])
+
+cc_library(
+    name = "acero",
+    srcs = glob(
+        [
+            "*.cc",
+        ],
+        exclude = [
+            "test_*.cc",
+            "*_test.cc",
+            "benchmark_*.cc",
+            "*_benchmark.cc",
+            "generate_fuzz_corpus.cc",
+            "generate_tensor_fuzz_corpus.cc",
+            "*_fuzz.cc",
+            "file_to_stream.cc",
+            "stream_to_file.cc",
+        ],
+    ),
+    hdrs = glob(
+        [
+            "*.h",
+        ],
+        exclude = [
+            "test_*.h",
+            "benchmark_*.h",
+        ],
+    ),
+    include_prefix = "arrow/acero",
+    strip_include_prefix = "",
+    copts = [
+        # "-DARROW_HAVE_RUNTIME_BMI2",
+        # "-mbmi2",
+        # "-DARROW_HAVE_RUNTIME_AVX2",
+        # "-mavx2",
+        # "-DARROW_HAVE_RUNTIME_AVX512",
+        # "-mavx512f",
+        # "-mavx512cd",
+        # "-mavx512bw",
+        # "-mavx512dq",
+        # "-mavx512vl",
+    ],
+    deps = [
+        "//cpp/src/arrow:core",
+        "//cpp/src/arrow/ipc:ipc",
+        "//cpp/src/generated:generated",
+        "@com_github_google_flatbuffers//:flatbuffers",
+    ],
+)
diff --git a/cpp/src/arrow/csv/BUILD.bazel b/cpp/src/arrow/csv/BUILD.bazel
new file mode 100644
index 0000000000..6b8a630fcb
--- /dev/null
+++ b/cpp/src/arrow/csv/BUILD.bazel
@@ -0,0 +1,32 @@
+package(default_visibility = ["//visibility:public"])
+
+cc_library(
+    name = "csv",
+    srcs = glob(
+        [
+            "*.cc",
+        ],
+        exclude = [
+            "test_*.cc",
+            "*_test.cc",
+            "*_benchmark.cc",
+        ],
+    ),
+    hdrs = glob(
+        [
+            "*.h",
+        ],
+        exclude = [
+            "test_*.h",
+        ],
+    ),
+    include_prefix = "arrow/csv",
+    strip_include_prefix = "",
+    deps = [
+        "//cpp/src/arrow:core",
+        "//cpp/src/arrow/ipc:ipc",
+        # "//cpp/src/generated:generated",
+        # "@com_github_google_flatbuffers//:flatbuffers",
+        # ":feather_fbs",
+    ],
+)
diff --git a/cpp/src/arrow/dataset/BUILD.bazel b/cpp/src/arrow/dataset/BUILD.bazel
new file mode 100644
index 0000000000..03673a1469
--- /dev/null
+++ b/cpp/src/arrow/dataset/BUILD.bazel
@@ -0,0 +1,45 @@
+package(default_visibility = ["//visibility:public"])
+
+cc_library(
+    name = "dataset",
+    srcs = glob(
+        [
+            "*.cc",
+        ],
+        exclude = [
+            "test_*.cc",
+            "*_test.cc",
+            "benchmark_*.cc",
+            "*_benchmark.cc",
+        ],
+    ),
+    hdrs = glob(
+        [
+            "*.h",
+        ],
+        exclude = [
+            "test_*.h",
+            "benchmark_*.h",
+        ],
+    ),
+    include_prefix = "arrow/dataset",
+    strip_include_prefix = "",
+    copts = [
+        # "-DARROW_HAVE_RUNTIME_BMI2",
+        # "-mbmi2",
+        # "-DARROW_HAVE_RUNTIME_AVX2",
+        # "-mavx2",
+        # "-DARROW_HAVE_RUNTIME_AVX512",
+        # "-mavx512f",
+        # "-mavx512cd",
+        # "-mavx512bw",
+        # "-mavx512dq",
+        # "-mavx512vl",
+    ],
+    deps = [
+        "//cpp/src/arrow:core",
+        "//cpp/src/arrow/ipc:ipc",
+        "//cpp/src/generated:generated",
+        "@com_github_google_flatbuffers//:flatbuffers",
+    ],
+)
diff --git a/cpp/src/arrow/filesystem/BUILD.bazel b/cpp/src/arrow/filesystem/BUILD.bazel
new file mode 100644
index 0000000000..b23a4dc587
--- /dev/null
+++ b/cpp/src/arrow/filesystem/BUILD.bazel
@@ -0,0 +1,55 @@
+package(default_visibility = ["//visibility:public"])
+
+cc_library(
+    name = "filesystem",
+    srcs = glob(
+        [
+            "*.cc",
+        ],
+        exclude = [
+            "test_*.cc",
+            "*_test.cc",
+            "benchmark_*.cc",
+            "*_benchmark.cc",
+            "azurefs*.cc",
+            "gcsfs*.cc",
+            "hdfs*.cc",
+            "s3fs*.cc",
+            "s3*.cc",
+        ],
+    ),
+    hdrs = glob(
+        [
+            "*.h",
+        ],
+        exclude = [
+            "test_*.h",
+            "benchmark_*.h",
+            "azurefs*.h",
+            "gcsfs*.h",
+            "hdfs*.h",
+            "s3fs*.h",
+            "s3*.h",
+        ],
+    ),
+    copts = [
+        # "-DARROW_HAVE_RUNTIME_BMI2",
+        # "-mbmi2",
+        # "-DARROW_HAVE_RUNTIME_AVX2",
+        # "-mavx2",
+        # "-DARROW_HAVE_RUNTIME_AVX512",
+        # "-mavx512f",
+        # "-mavx512cd",
+        # "-mavx512bw",
+        # "-mavx512dq",
+        # "-mavx512vl",
+    ],
+    include_prefix = "arrow/filesystem",
+    strip_include_prefix = "",
+    deps = [
+        "//cpp/src/arrow:core",
+        "//cpp/src/arrow/ipc",
+        "//cpp/src/generated",
+        "@com_github_google_flatbuffers//:flatbuffers",
+    ],
+)
diff --git a/cpp/src/arrow/flight/BUILD.bazel b/cpp/src/arrow/flight/BUILD.bazel
new file mode 100644
index 0000000000..93d4c1a269
--- /dev/null
+++ b/cpp/src/arrow/flight/BUILD.bazel
@@ -0,0 +1,58 @@
+package(default_visibility = ["//visibility:public"])
+
+cc_library(
+    name = "flight_grpc_srcs",
+    textual_hdrs = ["//format:flight_grpc_pb_srcs"],  # ← This is critical!
+)
+
+cc_library(
+    name = "flight",
+    srcs = glob(
+        [
+            "*.cc",
+            "transport/grpc/*.cc",
+        ],
+        exclude = [
+            "**/*_test.cc",
+            "**/test_*.cc",
+            "**/*_benchmark.cc",
+            "**/benchmark_*.cc",
+            "**/perf_*.cc",
+        ],
+    ),
+    hdrs = glob(
+        [
+            "*.h",
+            "transport/grpc/*.h",
+        ],
+        exclude = [
+            "**/*_test.h",
+            "**/test_*.h",
+        ],
+    ),
+    # copts = [
+    #     "-DARROW_HAVE_RUNTIME_BMI2",
+    #     "-mbmi2",
+    #     "-DARROW_HAVE_RUNTIME_AVX2",
+    #     "-mavx2",
+    #     "-DARROW_HAVE_RUNTIME_AVX512",
+    #     "-mavx512f",
+    #     "-mavx512cd",
+    #     "-mavx512bw",
+    #     "-mavx512dq",
+    #     "-mavx512vl",
+    # ],
+    include_prefix = "arrow/flight",
+    strip_include_prefix = "",
+    deps = [
+        "//cpp/src/arrow:core",
+        "//cpp/src/arrow/ipc:ipc",
+        # "//format:flight_cc_proto",
+        "//format:flight_grpc",
+        ":flight_grpc_srcs",
+        "@grpc//:grpc++",
+        "@opentelemetry-cpp//api:api",
+    ],
+       # includes = ["."],
+    # textual_hdrs = [":flight_grpc"],
+)
diff --git a/cpp/src/arrow/ipc/BUILD.bazel b/cpp/src/arrow/ipc/BUILD.bazel
new file mode 100644
index 0000000000..3ce3f939d8
--- /dev/null
+++ b/cpp/src/arrow/ipc/BUILD.bazel
@@ -0,0 +1,69 @@
+package(default_visibility = ["//visibility:public"])
+
+
+# flatbuffer_cc_library(
+#     name = "ipc_message_fbs",
+#     srcs = ["feather.fbs"],
+#     outs = ["Message_generated.h"],
+#     include_prefix = "arrow/ipc",
+#     flatc_args = ["--gen-mutable", "--gen-object-api"],
+# )
+
+
+load("@com_github_google_flatbuffers//:build_defs.bzl", "flatbuffer_cc_library")
+
+flatbuffer_cc_library(
+    name = "feather_fbs",
+    srcs = [
+        "feather.fbs",
+    ],
+    out_prefix = "generated/",
+    cc_include_paths = ["."],
+)
+
+
+genrule(
+    name = "feather_fbs_gen",
+    srcs = ["feather.fbs"],
+    outs = ["Message_generated.h"],
+    tools = ["@flatbuffers//:flatc"],
+    cmd = "$(location @flatbuffers//:flatc) --cpp --gen-mutable --scoped-enums -o $(@D) $(location feather.fbs)",
+    visibility = ["//visibility:public"],
+)
+
+cc_library(
+    name = "ipc",
+    srcs = glob(
+        [
+            "*.cc",
+        ],
+        exclude = [
+            "test_*.cc",
+            "*_test.cc",
+            "*_benchmark.cc",
+            "generate_fuzz_corpus.cc",
+            "generate_tensor_fuzz_corpus.cc",
+            "*_fuzz.cc",
+            "file_to_stream.cc",
+            "stream_to_file.cc",
+        ],
+    ),
+    hdrs = glob(
+        [
+            "*.h",
+        ],
+        exclude = [
+            "test_*.h",
+        ],
+    ),
+    include_prefix = "arrow/ipc",
+    strip_include_prefix = "",
+    visibility = ["//visibility:public"],
+    deps = [
+        "//cpp/src/arrow:core",
+        # "//format:fbs",
+        "//cpp/src/generated:generated",
+        "@com_github_google_flatbuffers//:flatbuffers",
+        ":feather_fbs",
+    ],
+)
diff --git a/cpp/src/generated/BUILD.bazel b/cpp/src/generated/BUILD.bazel
new file mode 100644
index 0000000000..f10052e0b2
--- /dev/null
+++ b/cpp/src/generated/BUILD.bazel
@@ -0,0 +1,11 @@
+package(default_visibility = ["//visibility:public"])
+
+
+cc_library(
+    name = "generated",
+    srcs = glob(["parquet_types.cpp"]),
+    hdrs = glob(["*_generated.h", "parquet_types.h", "parquet_types.tcc"]),
+    deps = ["@thrift", "//cpp/src/arrow:core", "//cpp/src/parquet:windows_h"],
+    include_prefix = "generated",
+    strip_include_prefix = "",
+)
diff --git a/cpp/src/parquet/BUILD.bazel b/cpp/src/parquet/BUILD.bazel
new file mode 100644
index 0000000000..936dd2a9bb
--- /dev/null
+++ b/cpp/src/parquet/BUILD.bazel
@@ -0,0 +1,102 @@
+package(default_visibility = ["//visibility:public"])
+
+
+# Generate the header file, parquet_version.h
+genrule(
+    name = "parquet_version_h",
+    srcs = ["parquet_version.h.in"],
+    outs = ["parquet_version.h"],
+    cmd = """
+        # Start with the source file
+        cat $(location parquet_version.h.in) > $@
+
+        # Perform @VAR@ substitutions
+        sed -i'' \
+            -e 's/@ARROW_VERSION_MAJOR@/21/g' \
+            -e 's/@ARROW_VERSION_MINOR@/0/g' \
+            -e 's/@ARROW_VERSION_PATCH@/0/g' \
+            -e 's/@ARROW_SO_VERSION@/2100/g' \
+            -e 's/@ARROW_FULL_SO_VERSION@/2100.0.0/g' \
+            -e 's/@ARROW_VERSION@/21.0.0/g' \
+            $@
+    """,
+)
+
+cc_library(
+    name = "windows_h",
+    hdrs = ["windows_compatibility.h", "windows_fixup.h"],
+    include_prefix = "parquet",
+    strip_include_prefix = "",
+)
+
+
+
+cc_library(
+    name = "parquet",
+    srcs = glob(
+        [
+            "*.cc",
+            "arrow/*.cc",
+            "encryption/*.cc",
+            "geospatial/*.cc",
+        ],
+        exclude = [
+            "**/*_test.cc",
+            "**/test_*.cc",
+            "arrow/generate_fuzz_corpus.cc",
+            "util/*_test.cc",
+            "util/*simd512*.cc",
+            "util/*avx512*.cc",
+            "**/*_benchmark.cc",
+            "**/benchmark_*.cc",
+            "encryption/encryption_internal_nossl.cc",
+        ],
+    ),
+    hdrs = glob(
+        [
+            "*.h",
+            "api/*.h",
+            "arrow/*.h",
+            "encryption/*.h",
+            "geospatial/*.h",
+        ],
+        exclude = [
+            "**/*_test.h",
+            "**/test_*.h",
+            "util/*simd512*.h",
+            "util/*avx512*.h",
+            "io/hdfs*.h",
+        ],
+    ) + [":parquet_version_h"],
+    copts = [
+        # "-DARROW_HAVE_RUNTIME_BMI2",
+        # "-mbmi2",
+        # "-DARROW_HAVE_RUNTIME_AVX2",
+        # "-mavx2",
+        # "-DARROW_HAVE_RUNTIME_AVX512",
+        # "-mavx512f",
+        # "-mavx512cd",
+        # "-mavx512bw",
+        # "-mavx512dq",
+        # "-mavx512vl",
+    ],
+    include_prefix = "parquet",
+    strip_include_prefix = "",
+    deps = [
+        "@jemalloc",
+        "@opentelemetry-cpp//exporters/ostream:ostream_span_exporter",
+        "@opentelemetry-cpp//exporters/ostream:ostream_log_record_exporter",
+        "@opentelemetry-cpp//exporters/otlp:otlp_http_exporter",
+        "@opentelemetry-cpp//exporters/otlp:otlp_http_log_record_exporter",
+        "@opentelemetry-proto//:trace_service_proto_cc",
+        "@protobuf",
+        "@rapidjson",
+        "@snappy",
+        "@xsimd",
+        "@thrift",
+        "//cpp/src/arrow:core",
+        "//cpp/src/arrow/ipc:ipc",
+        "//cpp/src/arrow/filesystem:filesystem",
+        "//cpp/src/generated:generated",
+    ],
+)
diff --git a/format/BUILD.bazel b/format/BUILD.bazel
new file mode 100644
index 0000000000..ae18fa70ee
--- /dev/null
+++ b/format/BUILD.bazel
@@ -0,0 +1,79 @@
+package(default_visibility = ["//visibility:public"])
+
+load("@com_github_google_flatbuffers//:build_defs.bzl", "flatbuffer_cc_library")
+load("@rules_proto_grpc_cpp//:defs.bzl", "cpp_grpc_library")
+
+flatbuffer_cc_library(
+    name = "fbs",
+    srcs = [
+        "File.fbs",
+        "Message.fbs",
+        "Schema.fbs",
+        "SparseTensor.fbs",
+        "Tensor.fbs",
+    ],
+    out_prefix = "generated/",
+)
+
+### flight proto
+proto_library(
+    name = "flight_proto",
+    srcs = [
+        "Flight.proto",
+    ],
+    import_prefix = "arrow/flight",
+    strip_import_prefix = "",
+    deps = [
+        "@protobuf//:any_proto",
+        "@protobuf//:descriptor_proto",
+        "@protobuf//:timestamp_proto",
+    ],
+)
+
+cc_proto_library(
+    name = "flight_cc_proto",
+    deps = [":flight_proto"],
+)
+
+cpp_grpc_library(
+    name = "flight_grpc",
+    protos = [":flight_proto"],
+)
+
+### flight sql proto
+proto_library(
+    name = "flight_sql_proto",
+    srcs = [
+        "FlightSql.proto",
+    ],
+    import_prefix = "arrow/flight/sql",
+    strip_import_prefix = "",
+    deps = [
+        "@protobuf//:any_proto",
+        "@protobuf//:descriptor_proto",
+        "@protobuf//:timestamp_proto",
+    ],
+)
+
+cc_proto_library(
+    name = "flight_sql_cc_proto",
+    deps = [":flight_sql_proto"],
+)
+
+### substrait proto
+proto_library(
+    name = "substrait_proto",
+    srcs = [
+        "cpp/proto/substrait/extension_rels.proto",
+    ],
+    strip_import_prefix = "cpp/proto",
+    deps = [
+        "@substrait//:algebra_proto",
+        "@substrait//:extended_expression_proto",
+    ],
+)
+
+cc_proto_library(
+    name = "substrait_pb",
+    deps = [":substrait_proto"],
+)
